File: src/swift/4_BehavioralPatterns/11_Template/TemplateMethod.swift
import Foundation
import XCTest

// no abstract classes, so we fail in abstract methods
class Game
{
  func run() // this is the template method
  {
    start()
    while !haveWinner
    {
      takeTurn()
    }
    print("Player \(winningPlayer) wins!")
  }

  internal func start()
  {
    precondition(false, "this method needs to be overridden")
  }

  internal var haveWinner: Bool
  {
    get {
      precondition(false, "this method needs to be overridden")
    }
  }

  internal func takeTurn()
  {
    precondition(false, "this method needs to be overridden")
  }

  internal var winningPlayer: Int
  {
    get {
      precondition(false, "this method needs to be overridden")
    }
  }

  internal var currentPlayer = 0
  internal let numberOfPlayers: Int

  init(_ numberOfPlayers: Int)
  {
    self.numberOfPlayers = numberOfPlayers
  }
}

// simulate a game of chess
class Chess : Game
{
  private let maxTurns = 10
  private var turn = 1

  init() {
    super.init(2)
  }

  override func start()
  {
    print("Starting a game of chess with \(numberOfPlayers) players.")
  }

  override var haveWinner: Bool
  {
    return turn == maxTurns
  }

  override func takeTurn()
  {
    print("Turn \(turn) taken by player \(currentPlayer).")
    currentPlayer = (currentPlayer + 1) % numberOfPlayers
    turn += 1
  }

  override var winningPlayer: Int
  {
    return currentPlayer
  }
}

func main()
{
  let chess = Chess()
  chess.run()
}

main()

File: src/csharp/4_BehavioralPatterns/11_Template/TemplateMethod.cs
using static System.Console;

namespace DotNetDesignPatternDemos.Behavioral.TemplateMethod
{
  public abstract class Game
  {
    public void Run()
    {
      Start();
      while (!HaveWinner)
        TakeTurn();
      WriteLine($"Player {WinningPlayer} wins.");
    }

    protected abstract void Start();
    protected abstract bool HaveWinner { get; }
    protected abstract void TakeTurn();
    protected abstract int WinningPlayer { get; }

    protected int currentPlayer;
    protected readonly int numberOfPlayers;

    public Game(int numberOfPlayers)
    {
      this.numberOfPlayers = numberOfPlayers;
    }
  }

  // simulate a game of chess
  public class Chess : Game
  {
    public Chess() : base(2)
    {
    }

    protected override void Start()
    {
      WriteLine($"Starting a game of chess with {numberOfPlayers} players.");
    }

    protected override bool HaveWinner => turn == maxTurns;

    protected override void TakeTurn()
    {
      WriteLine($"Turn {turn++} taken by player {currentPlayer}.");
      currentPlayer = (currentPlayer + 1) % numberOfPlayers;
    }

    protected override int WinningPlayer => currentPlayer;

    private int maxTurns = 10;
    private int turn = 1;
  }

  public class Demo
  {
    static void Main(string[] args)
    {
      var chess = new Chess();
      chess.Run();
    }
  }
}